File: leet/balancedBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/*
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {

  let result = true;

  const helper = (node) => {
      if(node === null){
          return 0;
      }

      if(node.left === null && node.right === null) {
          return 1;
      }

      const left_val = helper(node.left);
      const right_val = helper(node.right);

      if(Math.abs(left_val-right_val) > 1){
          result = false;
      }

      return (Math.max(left_val, right_val)+1);

  }
  helper(root);
  return result

};

File: leet/balancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.isBalancedHelper(root) != -1

    def isBalancedHelper(self, node):
        if not node:
            return 0

        left = self.isBalancedHelper(node.left)
        right = self.isBalancedHelper(node.right)

        if left == -1 or right == -1 or abs(left-right) > 1:
            return -1  # Houston, we have a balance problem
        return max(left, right) + 1
